### DISCUSSION TITLE
Kinematic Hardening Coreespondence Model Development in Peridigm

### DISCUSSION THREAD

**Discussion by ppandit95** [peridigm/peridigm#198] [General]
Dear @johntfoster and Peridigm Community

As I am trying to implement kinematic hardening model in pridigm for which as per the dependency graph of Isotropic Hardening I have made following new files 
1. kinematic_hardening_correspondence.cxx
2. kinematic_hardening_correspondence.h
3. Peridigm_KinematicHardeningPlasticCorrespondenceMaterial.hpp
4. Peridigm_KinematicHardeningPlasticCorrespondenceMaterial.cpp
 by modifying the relevant files of Isotropic hardening model as well as modified Peridigm_MaterialFactory.cpp to add a else case which calls the relevant class for the new material model but while building the peridigm I ran into error at 98% as 

`/usr/bin/ld: ../../libPeridigmLib.a(Peridigm_KinematicHardeningPlasticCorrespondenceMaterial.cpp.o): in function `PeridigmNS::KinematicHardeningPlasticCorrespondenceMaterial::computeCauchyStress(double, int, PeridigmNS::DataManager&) const':
Peridigm_KinematicHardeningPlasticCorrespondenceMaterial.cpp:(.text+0x34f): undefined reference to `void CORRESPONDENCE::updateElasticKinematicHardeningPlasticCauchyStress<double>(double const*, double const*, double const*, double*, double*, double const*, double*, int, double, double, double, double, bool, double, double, double, double, double, double)'
collect2: error: ld returned 1 exit status
make[2]: *** [src/compute/unit_test/CMakeFiles/utPeridigm_Compute_Linear_Momentum.dir/build.make:626: src/compute/unit_test/utPeridigm_Compute_Linear_Momentum] Error 1
/usr/bin/ld: ../../libPeridigmLib.a(Peridigm_KinematicHardeningPlasticCorrespondenceMaterial.cpp.o): in function `PeridigmNS::KinematicHardeningPlasticCorrespondenceMaterial::computeCauchyStress(double, int, PeridigmNS::DataManager&) const':
Peridigm_KinematicHardeningPlasticCorrespondenceMaterial.cpp:(.text+0x34f):make[1]: *** [CMakeFiles/Makefile2:1515: src/compute/unit_test/CMakeFiles/utPeridigm_Compute_Linear_Momentum.dir/all] Error 2
 undefined reference to `make[1]: *** Waiting for unfinished jobs....
void CORRESPONDENCE::updateElasticKinematicHardeningPlasticCauchyStress<double>(double const*, double const*, double const*, double*, double*, double const*, double*, int, double, double, double, double, bool, double, double, double, double, double, double)'
/usr/bin/ld: ../../libPeridigmLib.a(collect2: error: ld returned 1 exit status
Peridigm_KinematicHardeningPlasticCorrespondenceMaterial.cpp.o): in function `PeridigmNS::KinematicHardeningPlasticCorrespondenceMaterial::computeCauchyStress(double, int, PeridigmNS::DataManager&) const':
Peridigm_KinematicHardeningPlasticCorrespondenceMaterial.cpp:(.text+0x34f): undefined reference to `void CORRESPONDENCE::updateElasticKinematicHardeningPlasticCauchyStress<double>(double const*, double const*, double const*, double*, double*, double const*, double*, int, double, double, double, double, bool, double, double, double, double, double, double)'
collect2: error: ld returned 1 exit status
make[2]: *** [src/compute/unit_test/CMakeFiles/utPeridigm_Compute_Angular_Momentum.dir/build.make:626: src/compute/unit_test/utPeridigm_Compute_Angular_Momentum] Error 1
make[2]: *** [src/compute/unit_test/CMakeFiles/utPeridigm_Compute_Force.dir/build.make:641: src/compute/unit_test/utPeridigm_Compute_Force] Error 1
make[1]: *** [CMakeFiles/Makefile2:1577: src/compute/unit_test/CMakeFiles/utPeridigm_Compute_Angular_Momentum.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:1608: src/compute/unit_test/CMakeFiles/utPeridigm_Compute_Force.dir/all] Error 2
/usr/bin/ld: libPeridigmLib.a(Peridigm_KinematicHardeningPlasticCorrespondenceMaterial.cpp.o): in function `PeridigmNS::KinematicHardeningPlasticCorrespondenceMaterial::computeCauchyStress(double, int, PeridigmNS::DataManager&) const':
Peridigm_KinematicHardeningPlasticCorrespondenceMaterial.cpp:(.text+0x34f): undefined reference to `void CORRESPONDENCE::updateElasticKinematicHardeningPlasticCauchyStress<double>(double const*, double const*, double const*, double*, double*, double const*, double*, int, double, double, double, double, bool, double, double, double, double, double, double)'
/usr/bin/ld: ../../collect2: error: ld returned 1 exit status
libPeridigmLib.a(Peridigm_KinematicHardeningPlasticCorrespondenceMaterial.cpp.o): in function `PeridigmNS::KinematicHardeningPlasticCorrespondenceMaterial::computeCauchyStress(double, int, PeridigmNS::DataManager&) const':
Peridigm_KinematicHardeningPlasticCorrespondenceMaterial.cpp:(.text+0x34f): undefined reference to `void CORRESPONDENCE::updateElasticKinematicHardeningPlasticCauchyStress<double>(double const*, double const*, double const*, double*, double*, double const*, double*, int, double, double, double, double, bool, double, double, double, double, double, double)'
collect2: error: ld returned 1 exit status
make[2]: *** [src/CMakeFiles/Peridigm.dir/build.make:625: src/Peridigm] Error 1
make[2]: *** [src/compute/unit_test/CMakeFiles/utPeridigm_Compute_Kinetic_Energy.dir/build.make:626: src/compute/unit_test/utPeridigm_Compute_Kinetic_Energy] Error 1
make[1]: *** [CMakeFiles/Makefile2:1457: src/CMakeFiles/Peridigm.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:1546: src/compute/unit_test/CMakeFiles/utPeridigm_Compute_Kinetic_Energy.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
`Moreover, the function for which undefined reference is shown is already defined in one of the files but the error still persists.
Any headers in what is going wrong here will prove quite beneficial .

Many Thanks
Pushkar

**Comment by smenon92** [comment]
Hi,
Did you edit the CmakeLists.txt file in the materials folder to include the new cxx file (kinematic_hardening_correspondence.cxx)?

**Reply by ppandit95** [reply]
Thanks, @smenon92 for pointing that out as I haven't edited that, and most probably I am getting the error due to that.....

**Reply by ppandit95** [reply]
I did modify the CMakeLists.txt and it compiled without error this time....

**Reply by smenon92** [reply]
Great

**Comment by ppandit95** [comment]
Thanks for helping me out here....

**Comment by ppandit95** [comment]
Dear @johntfoster and Peridigm Community,

I have implemented the Kinematic_Hardening_Model but ran into errors upon execution of Peridigm executable. So I wish to debug the error received as:-

../../../src/Peridigm 3block_isotropic.yaml 

-- Peridigm
-- version 1.5.0 (Dev)

[pushkar-GL63-8RE:32476] *** Process received signal ***
[pushkar-GL63-8RE:32476] Signal: Segmentation fault (11)
[pushkar-GL63-8RE:32476] Signal code: Address not mapped (1)
[pushkar-GL63-8RE:32476] Failing at address: (nil)
[pushkar-GL63-8RE:32476] [ 0] /usr/lib/x86_64-linux-gnu/libpthread.so.0(+0x153c0)[0x7f4f07e813c0]
[pushkar-GL63-8RE:32476] [ 1] ../../../src/Peridigm(_ZN10PeridigmNS15PeridigmFactory6createENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKP19ompi_communicator_tN7Teuchos3RCPINS_14DiscretizationEEEPd+0x433)[0x5613f9f27cb3]
[pushkar-GL63-8RE:32476] [ 2] ../../../src/Peridigm(_ZN10PeridigmNS15PeridigmFactory6createENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKP19ompi_communicator_tPd+0xdc)[0x5613f9f2753c]
[pushkar-GL63-8RE:32476] [ 3] ../../../src/Peridigm(main+0x1b1)[0x5613f9da0861]
[pushkar-GL63-8RE:32476] [ 4] /usr/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3)[0x7f4f07ca10b3]
[pushkar-GL63-8RE:32476] [ 5] ../../../src/Peridigm(_start+0x2e)[0x5613f9db8cee]
[pushkar-GL63-8RE:32476] *** End of error message ***
Segmentation fault (core dumped)


Any help in this regard will prove beneficial.

Many Thanks 
Pushkar

**Reply by oldninja** [reply]
@ppandit95 is hard to debug without knowing what was changed by adding the Kinematic_Hardening_Model. I'm not sure what would be the best way to review the proposed changes and recreate this issue. Do you have a personal repository of Peridigm to check there?

**Reply by ppandit95** [reply]
@oldninja thats an interesting suggestion but I did manage to run the model by editing CMakeLists.txt in materials folder.....

**Reply by jritta** [reply]
@ppandit95 Did you already tried to run it without MPI? Does this work? Does the input deck runs with the current master of Peridigm (by using an already implemented model)?
Best
Jonas
